functions/go/set-namespace/transformer/namespace.go

```diff
@@ -18,206 +18,311 @@ import (
 	"strings"
 
 	"github.com/GoogleContainerTools/kpt-functions-sdk/go/fn"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
-func SetNamespace(rl *fn.ResourceList) (bool, error) {
-	tc := NamespaceTransformer{}
+// Run provides the main workflow to update the ResourceList.Items "namespace" value.
+func Run(rl *fn.ResourceList) (bool, error) {
+	tc := SetNamespace{}
 	// Get "namespace" arguments from FunctionConfig
 	err := tc.Config(rl.FunctionConfig)
 	if err != nil {
-		rl.Results = append(rl.Results, fn.ErrorResult(err))
-		return false, nil
+		rl.Results = append(rl.Results, fn.ErrorConfigObjectResult(err, rl.FunctionConfig))
+		return true, nil
 	}
 	// Update "namespace" to the proper resources.
-	oldNamespaces := map[string]struct{}{}
-	count := tc.Transform(rl.Items, oldNamespaces)
-	result := tc.LogResults(rl, count, oldNamespaces)
-	rl.Results = append(rl.Results, result)
+	results := tc.Transform(rl.Items)
+	rl.Results = append(rl.Results, results...)
 	return true, nil
 }
 
-func (p *NamespaceTransformer) LogResults(rl *fn.ResourceList, count int, oldNamespaces map[string]struct{}) *fn.Result {
-	if len(p.Errors) != 0 {
-		errMsg := strings.Join(p.Errors, "\n")
-		return fn.GeneralResult(errMsg, fn.Error)
-	}
-	if dupErr := fn.CheckResourceDuplication(rl); dupErr != nil {
-		return fn.ErrorResult(dupErr)
-	}
-	if count == 0 {
-		msg := fmt.Sprintf("all namespaces are already %q. no value changed", p.NewNamespace)
-		return fn.GeneralResult(msg, fn.Info)
-	}
-
-	oldNss := []string{}
-	for oldNs := range oldNamespaces {
-		oldNss = append(oldNss, `"`+oldNs+`"`)
-	}
-	msg := fmt.Sprintf("namespace %v updated to %q, %d value(s) changed", strings.Join(oldNss, ","), p.NewNamespace, count)
-	return fn.GeneralResult(msg, fn.Info)
+// SetNamespace defines structs to parse KRM resource "SetNamespace" (the custom function config) and "ConfigMap" data.
+// it provides the method "Config" to read the function configs from ResourceList.FunctionConfig
+// it provides the method "Transform" to change the "namespace" and update the "config.kubernetes.io/depends-on" annotation.
+type SetNamespace struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
```

Comment:
protobuf tag is not useful here.
Instead of using `metav1`, we can use https://github.com/kubernetes-sigs/kustomize/blob/b727febd08b3902b959e48e2f2649c77a3ed911d/kyaml/yaml/types.go#L121-L152

Reply:
I'm trying to not import kyaml for now. Or we can define similar things on the sdk.

Reply:
`sdk/go/fn` depends on `kyaml`. I guess you mean you don't want to depend on `kyaml` explicitly.
One thing we can do here is to define types in https://github.com/kubernetes-sigs/kustomize/blob/b727febd08b3902b959e48e2f2649c77a3ed911d/kyaml/yaml/types.go#L121-L152 as alias in `sdk/go/fn`.

Types in `metav1` only has json tags but not yaml tags.

Reply:
Yea, I want to simplify the go dependencies as much as possible, otherwise users can easily encounter dependency version conflict. I think it's better to not depend on kyaml directly in functions, but either provide the needed struct in SDK (like you suggested ) or even add the struct inline.  I added the structure in this function with a TODO to let go/fn provides the meta types eventually.

Reply:
Great. Changing this removes another dependency metav1.
I also realized another potential risk condition, where that users may generate yaml serialized ConfigMap (as you mentioned in the other comment), does it mean if this ConfigMap cannot be parsed as below since the metav1 does not have yaml tag? Or did I miss anything? 
```yaml
var cm metav1.ConfigMap
fnConfigObject.As(&cm)
```

If this is a concern, then I guess we need to provide a yaml tag handler in the go/fn.

Reply:
`As` function in `go/fn` has considered it. It relies on json tags to deserialize yaml.
If a user tries to deserialize yaml using `go-yaml` or `kyaml` directly (e.g. deserialize yaml bytes to metav1.ConfigMap), he/she will run into issues.

When we define types in our projects, we should always try to add both json and yaml tags.

Reply:
Wonderful!  Didn't pay close attention to json/yaml go libraries before.